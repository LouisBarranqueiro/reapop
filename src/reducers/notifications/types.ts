import {POSITIONS, STATUSES} from '../../constants'

export interface NotificationButton {
    name: string
    primary?: boolean
    onClick?: (...args: any[]) => void
}

export type Status = typeof STATUSES[keyof typeof STATUSES]
export type Position = typeof POSITIONS[keyof typeof POSITIONS]

export interface Notification {
    id: string
    title?: string
    message?: string
    status: Status
    position: Position
    buttons: NotificationButton[]
    image?: string
    dismissAfter?: number
    dismissible?: boolean
    onAdd?: (...args: any[]) => void
    onDismiss?: (...args: any[]) => void
    showDismissButton?: boolean
    allowHTML?: boolean
    [index: string]: any
}

export type NewNotification = Partial<Notification>
