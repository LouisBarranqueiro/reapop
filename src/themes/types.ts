import {CSSProperties} from 'react'
import {Notification, Position} from '../reducers/notifications/types'

export type NotificationButtonState = {isHovered: boolean; isActive: boolean}

export type Theme = {
    container: (position: Position, singleContainer: boolean) => CSSProperties
    notificationDismissIcon: (notification: Notification) => CSSProperties
    notification: (notification: Notification) => CSSProperties
    notificationIcon: (notification: Notification) => CSSProperties
    notificationImageContainer: (notification: Notification) => CSSProperties
    notificationImage: (notification: Notification) => CSSProperties
    notificationMeta: (notification: Notification) => CSSProperties
    notificationTitle: (notification: Notification) => CSSProperties
    notificationMessage: (notification: Notification) => CSSProperties
    notificationButtons: (notification: Notification) => CSSProperties
    notificationButton: (notification: Notification, position: number, state: NotificationButtonState) => CSSProperties
    notificationButtonText: (
        notification: Notification,
        position: number,
        state: NotificationButtonState
    ) => CSSProperties
}
