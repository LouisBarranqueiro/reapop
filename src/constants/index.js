export const DEFAULT_STATUS = 'default';
export const INFO_STATUS = 'info';
export const SUCCESS_STATUS = 'success';
export const WARNING_STATUS = 'warning';
export const ERROR_STATUS = 'error';

export const STATUS = {
  default: DEFAULT_STATUS,
  info: INFO_STATUS,
  success: SUCCESS_STATUS,
  warning: WARNING_STATUS,
  error: ERROR_STATUS
};

export const TOP = 't';
export const TOP_CENTER = 'tc';
export const TOP_LEFT_POSITION = 'tl';
export const TOP_RIGHT_POSITION = 'tr';
export const BOTTOM = 'b';
export const BOTTOM_CENTER = 'bc';
export const BOTTOM_LEFT_POSITION = 'bl';
export const BOTTOM_RIGHT_POSITION = 'br';

export const POSITIONS = {
  top: TOP,
  topCenter: TOP_CENTER,
  topLeft: TOP_LEFT_POSITION,
  topRight: TOP_RIGHT_POSITION,
  bottom: BOTTOM,
  bottomCenter: BOTTOM_CENTER,
  bottomLeft: BOTTOM_LEFT_POSITION,
  bottomRight: BOTTOM_RIGHT_POSITION
};

// default value for notifications
export const DEFAULT_NOTIFICATION = {
  status: DEFAULT_STATUS,
  position: TOP_RIGHT_POSITION,
  dismissible: true,
  dismissAfter: 5000,
  allowHTML: false,
  closeButton: false
};
