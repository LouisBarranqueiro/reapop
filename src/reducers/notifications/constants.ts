export enum NotificationAction {
    UpsertNotification = 'reapop/upsertNotification',
    DismissNotification = 'reapop/dismissNotification',
    DismissNotifications = 'reapop/dismissNotifications',
}
