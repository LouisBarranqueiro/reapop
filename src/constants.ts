import {NewNotification, Notification, Position} from './reducers/notifications/types'

export type NotificationConfig = {
    defaultProps: Partial<Notification>
    customizeNotification: ((notification: NewNotification) => void) | null
    generateId: ((notification: NewNotification) => string) | null
}

export const CONFIG: NotificationConfig = {
    defaultProps: {},
    customizeNotification: null,
    generateId: null,
}

export const STATUSES: {
    none: 'none'
    info: 'info'
    success: 'success'
    loading: 'loading'
    warning: 'warning'
    error: 'error'
} = Object.freeze({
    none: 'none',
    info: 'info',
    success: 'success',
    loading: 'loading',
    warning: 'warning',
    error: 'error',
})

export const POSITIONS: {
    topCenter: 'top-center'
    topLeft: 'top-left'
    topRight: 'top-right'
    bottomCenter: 'bottom-center'
    bottomLeft: 'bottom-left'
    bottomRight: 'bottom-right'
} = Object.freeze({
    topCenter: 'top-center',
    topLeft: 'top-left',
    topRight: 'top-right',
    bottomCenter: 'bottom-center',
    bottomLeft: 'bottom-left',
    bottomRight: 'bottom-right',
})

const classnamePrefix = 'reapop__'

export const classnames = {
    container: (position: Position, singleContainer: boolean) => {
        const classes = [`${classnamePrefix}container`, `${classnamePrefix}container--${position}`]
        if (singleContainer) {
            classes.push(`${classnamePrefix}container--single`)
        }
        return classes
    },
    notification: (notification: Notification) => [
        `${classnamePrefix}notification`,
        `${classnamePrefix}notification--${notification.status}`,
        `${classnamePrefix}notification--buttons-${notification.buttons.length}`,
        notification.dismissible ? `${classnamePrefix}notification--dismissible` : '',
    ],
    notificationIcon: `${classnamePrefix}notification-icon`,
    notificationImageContainer: `${classnamePrefix}notification-image-container`,
    notificationImage: `${classnamePrefix}notification-image`,
    notificationDismissIcon: `${classnamePrefix}notification-dismiss-icon`,
    notificationMeta: `${classnamePrefix}notification-meta`,
    notificationTitle: `${classnamePrefix}notification-title`,
    notificationMessage: `${classnamePrefix}notification-message`,
    notificationButtons: `${classnamePrefix}notification-buttons`,
    notificationButton: `${classnamePrefix}notification-button`,
    notificationButtonText: `${classnamePrefix}notification-button-text`,
}
