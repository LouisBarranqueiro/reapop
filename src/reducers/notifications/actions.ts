import {NotificationAction} from './constants'
import {NewNotification, Notification, Status} from './types'
import {isObject} from '../../utils'
import {prepareNotification} from '../../services/notifications'

export type UpsertNotificationAction = {
    type: NotificationAction.UpsertNotification
    payload: Notification
}

export type DismissNotificationAction = {
    type: NotificationAction.DismissNotification
    payload: string
}

export type DismissNotificationsAction = {
    type: NotificationAction.DismissNotifications
}

function upsertNotification(notification: Partial<Notification>): UpsertNotificationAction
function upsertNotification(message: string, options?: Partial<Notification>): UpsertNotificationAction
function upsertNotification(message: string, status: Status, options?: Partial<Notification>): UpsertNotificationAction
function upsertNotification(...args: unknown[]): UpsertNotificationAction {
    const lastArg = args[args.length - 1]
    let notification: NewNotification = {}

    if (typeof args[0] === 'string') {
        notification.message = args[0] as string
    }
    if (typeof args[1] === 'string') {
        notification.status = args[1] as Status
    }
    if (isObject(lastArg)) {
        notification = {...notification, ...(lastArg as NewNotification)}
    }

    return {
        type: NotificationAction.UpsertNotification,
        payload: prepareNotification(notification as NewNotification | Notification),
    }
}

export const dismissNotification = (notificationId: string): DismissNotificationAction => ({
    type: NotificationAction.DismissNotification,
    payload: notificationId,
})

export const dismissNotifications = (): DismissNotificationsAction => ({
    type: NotificationAction.DismissNotifications,
})

export const notify = upsertNotification

export type NotificationActions = UpsertNotificationAction | DismissNotificationAction | DismissNotificationsAction
