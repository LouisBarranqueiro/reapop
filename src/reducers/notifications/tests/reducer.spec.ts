import notificationsReducer from '../reducer'
import {dismissNotification, dismissNotifications, notify} from '../actions'
import {POSITIONS, STATUSES} from '../../../constants'
import {Notification} from '../types'

describe('notificationsReducer', () => {
    let notification: Notification
    let reducer: ReturnType<typeof notificationsReducer>
    beforeEach(() => {
        reducer = notificationsReducer()
        notification = {
            id: '1',
            dismissible: false,
            position: POSITIONS.bottomCenter,
            status: STATUSES.none,
            buttons: [],
        }
    })

    describe('init redux action', () => {
        it.each([
            ['default state', undefined],
            ['filled state', [notification]],
        ])('should return notifications (%s)', (_, state) => {
            expect(reducer(state, {type: 'redux-init'})).toEqual(state || [])
        })
    })

    describe('UpsertNotification', () => {
        it('should add notification', () => {
            expect(reducer(undefined, notify({id: '1'}))).toMatchSnapshot()
        })

        it('should update notification', () => {
            const notification2: Notification = {
                id: '2',
                dismissible: false,
                position: POSITIONS.topLeft,
                status: STATUSES.error,
                buttons: [],
            }
            const initialState = [notification, notification2]

            notification.dismissible = true
            notification.allowHTML = true

            const action = notify({id: '1', dismissible: true})
            expect(reducer(initialState, action)).toMatchSnapshot()
        })
    })

    describe('DismissNotification', () => {
        it('should dismiss notification (empty initial state)', () => {
            expect(reducer([], dismissNotification(notification.id))).toMatchSnapshot()
        })

        it('should dismiss notification (filled initial state)', () => {
            expect(reducer([notification], dismissNotification(notification.id))).toMatchSnapshot()
        })
    })

    describe('DismissNotifications', () => {
        it.each([
            ['without notifications', undefined],
            ['with notifications', [notification]],
        ])('should dismiss notifications (%s)', (_, initialState) => {
            expect(reducer(initialState, dismissNotifications())).toEqual([])
        })
    })
})
