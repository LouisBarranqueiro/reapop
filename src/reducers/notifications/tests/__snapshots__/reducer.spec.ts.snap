// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notificationsReducer DismissNotification should dismiss notification (empty initial state) 1`] = `Array []`;

exports[`notificationsReducer DismissNotification should dismiss notification (filled initial state) 1`] = `Array []`;

exports[`notificationsReducer UpsertNotification should add notification 1`] = `
Array [
  Object {
    "buttons": Array [],
    "id": "1",
    "status": "none",
  },
]
`;

exports[`notificationsReducer UpsertNotification should update notification 1`] = `
Array [
  Object {
    "buttons": Array [],
    "dismissible": true,
    "id": "1",
    "status": "none",
  },
  Object {
    "buttons": Array [],
    "dismissible": false,
    "id": "2",
    "position": "top-left",
    "status": "error",
  },
]
`;
